File: src/algebra1_one_step_equations_hard.py
"""
algebra1 - one_step_equations (hard)
Generated: 2026-02-22T03:47:35.209029
"""

from problem_utils import *

def generate():
    # For HARDER one-step equations (1600-1900), we need to push beyond typical one-step problems
    # Options:
    # 1. Fractional coefficients with complex numbers
    # 2. Equations requiring careful manipulation of radicals
    # 3. One-step but with algebraic expressions as coefficients
    # 4. Equations involving absolute values (technically one-step after setup)
    # 5. Rational equations that simplify to one-step
    
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Fractional coefficient with nested fractions
        num1 = nonzero(-7, 7)
        den1 = choice([2, 3, 4, 5, 6])
        num2 = nonzero(-9, 9)
        den2 = choice([2, 3, 4, 5, 6])
        
        # Equation: (num1/den1)x = num2/den2
        # Solution: x = (num2/den2) / (num1/den1) = (num2 * den1) / (den2 * num1)
        
        coeff = Rational(num1, den1)
        rhs = Rational(num2, den2)
        ans = rhs / coeff
        
        question = f"\\frac{{{num1}}}{{{den1}}}x = \\frac{{{num2}}}{{{den2}}}"
        
        solution = steps(
            f"Divide both sides by $\\frac{{{num1}}}{{{den1}}}$",
            f"$x = \\frac{{{num2}}}{{{den2}}} \\div \\frac{{{num1}}}{{{den1}}}$",
            f"$x = \\frac{{{num2}}}{{{den2}}} \\cdot \\frac{{{den1}}}{{{num1}}}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1600, 1750),
            topic="algebra1/one_step_equations",
            solution=solution
        )
    
    elif problem_type == 2:
        # Equation with radicals: sqrt(a) * x = b
        a_squared = choice([4, 9, 16, 25, 36, 49])
        a_val = sqrt(a_squared)
        b_val = nonzero(-10, 10)
        
        ans = Rational(b_val, int(a_val))
        
        question = f"\\sqrt{{{a_squared}}}x = {b_val}"
        
        solution = steps(
            f"Simplify: $\\sqrt{{{a_squared}}} = {int(a_val)}$",
            f"${int(a_val)}x = {b_val}$",
            f"Divide both sides by ${int(a_val)}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra1/one_step_equations",
            solution=solution
        )
    
    elif problem_type == 3:
        # Rational equation that's one-step: x/a = b/c
        a_val = nonzero(-8, 8)
        num_b = nonzero(-9, 9)
        den_c = choice([2, 3, 4, 5, 6, 7])
        
        # x/a = num_b/den_c
        # x = a * num_b / den_c
        
        ans = Rational(a_val * num_b, den_c)
        
        question = f"\\frac{{x}}{{{a_val}}} = \\frac{{{num_b}}}{{{den_c}}}"
        
        solution = steps(
            f"Multiply both sides by ${a_val}$",
            f"$x = {a_val} \\cdot \\frac{{{num_b}}}{{{den_c}}}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1650, 1800),
            topic="algebra1/one_step_equations",
            solution=solution
        )
    
    elif problem_type == 4:
        # Decimal coefficient requiring careful division
        decimals = [0.25, 0.5, 0.75, 1.25, 1.5, 2.5, 3.5]
        coeff_dec = choice(decimals)
        rhs_val = nonzero(-15, 15)
        
        # Convert to rational for exact answer
        coeff_rat = Rational(int(coeff_dec * 4), 4)
        ans = Rational(rhs_val) / coeff_rat
        
        question = f"{coeff_dec}x = {rhs_val}"
        
        solution = steps(
            f"Divide both sides by ${coeff_dec}$",
            f"$x = \\frac{{{rhs_val}}}{{{coeff_dec}}}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1600, 1750),
            topic="algebra1/one_step_equations",
            solution=solution
        )
    
    else:  # problem_type == 5
        # Complex rational equation: (ax + b)/c = d, solve for x (still one-step algebraically)
        # Wait, that's two-step. Let me do: x/(a/b) = c
        
        num_a = nonzero(-6, 6)
        den_b = choice([2, 3, 4, 5])
        c_val = nonzero(-10, 10)
        
        # x / (a/b) = c
        # x = c * (a/b) = ca/b
        
        coeff_frac = Rational(num_a, den_b)
        ans = c_val * coeff_frac
        
        question = f"\\frac{{x}}{{\\frac{{{num_a}}}{{{den_b}}}}} = {c_val}"
        
        solution = steps(
            f"Multiply both sides by $\\frac{{{num_a}}}{{{den_b}}}$",
            f"$x = {c_val} \\cdot \\frac{{{num_a}}}{{{den_b}}}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1700, 1850),
            topic="algebra1/one_step_equations",
            solution=solution
        )

emit(generate())

File: src/algebra1_one_step_equations_hard.jl
# algebra1 - one_step_equations (hard)
# Generated: 2026-03-08T19:52:23.236896

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra1/one_step_equations")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Fractional coefficient with large denominators (1800-2000)
        ans = randint(-50, 50)
        num = nonzero(-15, 15)
        den = choice([7, 9, 11, 13, 17, 19, 23])
        rhs = randint(-100, 100)
        coeff = num // den
        lhs_val = coeff * ans
        
        question_str = "Solve \\\$\\frac{$(num)}{$(den)}x = $(tex(lhs_val))\\\$"
        
        problem(
            question=question_str,
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                sol("Given", (num//den)*x ~ lhs_val),
                "Multiply both sides by \\\$\\frac{$(den)}{$(num)}\\\$",
                sol("Simplify", x ~ ans),
                sol("Answer", ans)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Equation with nested radicals (1900-2100)
        ans = choice([4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144])
        inner = randint(2, 8)
        k = nonzero(-12, 12)
        
        outer_val = inner * sqrt(ans) + k
        
        question_str = "Solve \\\$\\sqrt{x} = \\frac{$(tex(outer_val)) - $(k)}{$(inner)}\\\$"
        
        problem(
            question=question_str,
            answer=ans,
            difficulty=(1900, 2100),
            solution=steps(
                "Given: \\\$\\sqrt{x} = \\frac{$(tex(outer_val)) - $(k)}{$(inner)}\\\$",
                "Simplify right side: \\\$\\sqrt{x} = $(tex(sqrt(ans)))\\\$",
                "Square both sides",
                sol("Answer", ans)
            ),
            time=150
        )
        
    elseif problem_type == 3
        # Absolute value with fractional result (2000-2200)
        ans_num = randint(-30, 30)
        ans_den = choice([3, 5, 7, 9, 11])
        ans = ans_num // ans_den
        abs_ans = abs(ans)
        
        k = nonzero(-20, 20)
        rhs = abs_ans + k
        
        question_str = "Solve \\\$|x| + $(k) = $(tex(rhs))\\\$"
        
        ans_set = fmt_set(Set([ans, -ans]))
        
        problem(
            question=question_str,
            answer=ans_set,
            answer_type="set",
            difficulty=(2000, 2200),
            solution=steps(
                sol("Given", "|x| + $(k)" ~ rhs),
                "Subtract $(k) from both sides: \\\$|x| = $(tex(abs_ans))\\\$",
                "Absolute value gives two solutions",
                sol("Answer", ans_set)
            ),
            time=180
        )
        
    elseif problem_type == 4
        # Exponential equation requiring logarithm insight (2100-2300)
        base = choice([2, 3, 5, 7])
        ans = randint(4, 12)
        value = base^ans
        
        question_str = "Solve \\\$$(base)^x = $(value)\\\$"
        
        problem(
            question=question_str,
            answer=ans,
            difficulty=(2100, 2300),
            solution=steps(
                sol("Given", "$(base)^x" ~ value),
                "Recognize that \\\$$(value) = $(base)^{$(ans)}\\\$",
                "Since bases are equal, exponents must be equal",
                sol("Answer", ans)
            ),
            time=180
        )
        
    else
        # Rational equation with large coefficients (2200-2400)
        ans = nonzero(-25, 25)
        a = nonzero(-15, 15)
        b = nonzero(-30, 30)
        
        # a/(x) = b/ans, so a*ans = b*x
        # We want x = ans, so we need a*ans = b*ans, which means we set the RHS appropriately
        rhs_num = a * ans
        rhs_den = b
        
        question_str = "Solve \\\$\\frac{$(a)}{x} = \\frac{$(rhs_num)}{$(b*ans)}\\\$"
        
        problem(
            question=question_str,
            answer=ans,
            difficulty=(2200, 2400),
            solution=steps(
                sol("Given", "\\frac{$(a)}{x}" ~ "\\frac{$(rhs_num)}{$(b*ans)}"),
                "Cross multiply: \\\$$(a) \\cdot $(b*ans) = $(rhs_num) \\cdot x\\\$",
                "Simplify: \\\$$(a * b * ans) = $(rhs_num) x\\\$",
                "Divide both sides by $(rhs_num)",
                sol("Answer", ans)
            ),
            time=200
        )
    end
end